src/components/tl/ucp/reduce_scatter/reduce_scatter_knomial.c

```diff
@@ -0,0 +1,270 @@
+
+    local_seg_count = 0;
+    block_count     = ucc_sra_kn_compute_block_count(count, rank, p);
+    GOTO_PHASE(task->reduce_scatter_kn.phase);
+
+    if (KN_NODE_EXTRA == node_type) {
+        peer = ucc_knomial_pattern_get_proxy(p, rank);
+        UCPCHECK_GOTO(
+            ucc_tl_ucp_send_nb(sbuf, data_size, mem_type, peer, team, task),
+            task, out);
+    }
+
+    if (KN_NODE_PROXY == node_type) {
+        peer = ucc_knomial_pattern_get_extra(p, rank);
+        UCPCHECK_GOTO(
+            ucc_tl_ucp_recv_nb(scratch, data_size, mem_type, peer, team, task),
+            task, out);
+    }
+
+PHASE_EXTRA:
+    if ((KN_NODE_PROXY == node_type) || (KN_NODE_EXTRA == node_type)) {
+        if (UCC_INPROGRESS == ucc_tl_ucp_test(task)) {
+            SAVE_STATE(PHASE_EXTRA);
+            return task->super.super.status;
+        }
+        if (KN_NODE_EXTRA == node_type) {
+            goto out;
+        } else {
+            if (UCC_OK != (status = ucc_dt_reduce(sbuf, scratch, rbuf,
+                                                  count, dt, mem_type, &task->args))) {
+                tl_error(UCC_TL_TEAM_LIB(task->team),
+                         "failed to perform dt reduction");
+                task->super.super.status = status;
+                return status;
+            }
+            task->args.src.info.buffer = task->args.dst.info.buffer;
+        }
+    }
+    while(!ucc_knomial_pattern_loop_done(p)) {
+        step_radix  = ucc_sra_kn_compute_step_radix(rank, size, p);
+        block_count = ucc_sra_kn_compute_block_count(count, rank, p);
+        sbuf = (p->iteration == 0) ? task->args.src.info.buffer:
+            task->reduce_scatter_kn.scratch;
+        for (loop_step = 1; loop_step < radix; loop_step++) {
+            peer = ucc_knomial_pattern_get_loop_peer(p, rank, size, loop_step);
+            if (peer == UCC_KN_PEER_NULL)
+                continue;
+
+            peer_seg_index  = ucc_sra_kn_compute_seg_index(peer, p->radix_pow, p);
+            peer_seg_count  = ucc_sra_kn_compute_seg_size(block_count, step_radix, peer_seg_index);
+            peer_seg_offset = ucc_sra_kn_compute_seg_offset(block_count, step_radix, peer_seg_index);
+            UCPCHECK_GOTO(
+                ucc_tl_ucp_send_nb(PTR_OFFSET(sbuf, peer_seg_offset*dt_size),
+                                   peer_seg_count*dt_size, mem_type, peer, team, task),
+                task, out);
+        }
+
+        local_seg_index  = ucc_sra_kn_compute_seg_index(rank, p->radix_pow, p);
+        local_seg_count  = ucc_sra_kn_compute_seg_size(block_count, step_radix, local_seg_index);
+
+        rbuf = task->reduce_scatter_kn.scratch;
+        if (p->iteration != 0) {
+            rbuf = PTR_OFFSET(rbuf, block_count*dt_size);
+        }
+        for (loop_step = 1; loop_step < radix; loop_step++) {
+            peer = ucc_knomial_pattern_get_loop_peer(p, rank, size, loop_step);
+            if (peer == UCC_KN_PEER_NULL)
+                continue;
```

Comment:
can we do break instead?

Reply:
no, the peer can be non-null at some further loop_step